pkg/kvstore/etcd.go

```diff
@@ -1716,6 +1716,49 @@ func (e *etcdClient) ListAndWatch(ctx context.Context, name, prefix string, chan
 	return w
 }
 
+// UserEnforcePresence creates a user in etcd if not already present, and grants the specified roles.
+func (e *etcdClient) UserEnforcePresence(ctx context.Context, name string, roles []string) error {
+	scopedLog := e.getLogger().WithField(FieldUser, name)
+
+	scopedLog.Debug("Creating user")
+	_, err := e.client.Auth.UserAddWithOptions(ctx, name, "", &client.UserAddOptions{NoPassword: true})
+	if err != nil {
+		if errors.Is(err, v3rpcErrors.ErrUserAlreadyExist) {
+			scopedLog.Debug("User already exists")
+		} else {
+			return err
+		}
+	}
+
+	for _, role := range roles {
+		scopedLog.WithField(FieldRole, role).Debug("Granting role to user")
+
+		_, err := e.client.Auth.UserGrantRole(ctx, name, role)
```

Comment:
Does this return an error if the user already has this role?

Reply:
It does not: https://github.com/etcd-io/etcd/blob/main/server/auth/store.go#L542-L551